### PR TITLE
Threaded PKey Tracking

### PR DESCRIPTION
**Pull request by mgdickerson**
Added a map and mutex for tracking previously disabled pkey information for a given thread.

### CODE REVIEW

File: compiler-rt/lib/mpk_untrusted/alloc_site_handler.h [hunk 1]
```diff
@@ -104,6 +113,10 @@ class AllocSiteHandler {
   std::set<AllocSite> fault_set;
   // Thread safety mutex
   std::mutex mx;
+  // Mapping of thread-id to saved pkey
+  std::map<pid_t, PKeyInfo> pid_key_map;
```
Comment (ilovepi) [inline compiler-rt/lib/mpk_untrusted/alloc_site_handler.h]: Might make sense to name this something like `pkey_by_pid_map`, and then associated functions can be things like `get_pkey_by_pid(pid)`.  

The other alternative is to use a structure to track the metadata instead e.g., PKeyMetadata or whatever.

Note: I have no preference on snake vs camel case, its just what I typed, so use whatever you'd like

File: compiler-rt/lib/mpk_untrusted/alloc_site_handler.h [hunk 2]
```diff
@@ -193,6 +206,27 @@ class AllocSiteHandler {
...
+    // Obtain map key
+    const std::lock_guard<std::mutex> lock(px);
+
+    auto iter = pid_key_map.find(threadID);
+    // If PID does not contain key in map, return None.
+    if (iter == pid_key_map.end())
+      return llvm::None;
+
+    auto ret_val = iter->second;
+    pid_key_map.erase(threadID);
```
Comment (ilovepi) [inline compiler-rt/lib/mpk_untrusted/alloc_site_handler.h]: why remove from the map? getX() APIs don't usually destroy the data they retrieve.

File: compiler-rt/lib/mpk_untrusted/mpk_fault_handler.cpp [hunk 3]
```diff
@@ -53,22 +53,22 @@ void disablePageMPK(siginfo_t *si, void *arg) {
...
+  handler->storePidKey(getpid(), pkey_info);
   __mpk_untrusted::pkey_set(pkru_ptr, pkey, PKEY_ENABLE_ACCESS);
 
   __sanitizer::Report("INFO : Pkey(%d) has been set to ENABLE_ACCESS to enable "
                       "instruction access.\n",
                       pkey);
 }
 
-void enableThreadMPK(void *arg, uint32_t pkey) {
+void enableThreadMPK(void *arg, __mpk_untrusted::PKeyInfo pkey_info) {
```
Comment (ilovepi) [inline compiler-rt/lib/mpk_untrusted/mpk_fault_handler.cpp]: Just a general comment. since you're basically operating within a private namespace and in a cpp file, it is fine to use a `using` declaration to avoid what are normally unambiguous namespace prefixes. It would also be perfectly valid to put your code within the` __mpk_untrusted` namespace. `using` declarations are almost always OK in implementations, but should pretty much never happen in headers. being explicit is fine, but not strictly necessary.

File: compiler-rt/lib/mpk_untrusted/mpk_fault_handler.cpp [hunk 4]
```diff
@@ -53,22 +53,22 @@ void disablePageMPK(siginfo_t *si, void *arg) {
...
   uint32_t *pkru_ptr = __mpk_untrusted::pkru_ptr(arg);
-  __mpk_untrusted::pkey_set(pkru_ptr, last_pkey, last_access_rights);
-  __sanitizer::Report("INFO : Pkey(%d) has been reset to %d.\n", last_pkey,
-                      last_access_rights);
-  last_pkey = INVALID_PKEY;
-  last_access_rights = PKEY_ENABLE_ACCESS;
+  __mpk_untrusted::pkey_set(pkru_ptr, pkey_info.pkey, pkey_info.access_rights);
+  __sanitizer::Report("INFO : Pkey(%d) has been reset to %d.\n", pkey_info.pkey,
```
Comment (ilovepi) [inline compiler-rt/lib/mpk_untrusted/mpk_fault_handler.cpp]: can we change these calls to `Report` to a macro that is only defined when a compile time flag is set? This can be in another PR, but is something we should address soon.

File: compiler-rt/lib/mpk_untrusted/mpk_fault_handler.cpp [hunk 5]
```diff
@@ -90,7 +90,10 @@ void disableMPK(siginfo_t *si, void *arg) {
 void stepMPKHandle(int sig, siginfo_t *si, void *arg) {
   __sanitizer::Report(
       "Reached signal handler after single instruction step.\n");
-  enableThreadMPK(arg, si->si_pkey);
+  auto handler = __mpk_untrusted::AllocSiteHandler::get();
```
Comment (ilovepi) [inline compiler-rt/lib/mpk_untrusted/mpk_fault_handler.cpp]: Looking at the `get()` API here, it should probably be `GetOrInit()`, so that it is clear what is going on. This is unrelated to this PR, but would be good to fix.

File: compiler-rt/lib/mpk_untrusted/alloc_site_handler.h [hunk 6]
```diff
@@ -104,6 +113,10 @@ class AllocSiteHandler {
   std::set<AllocSite> fault_set;
   // Thread safety mutex
   std::mutex mx;
+  // Mapping of thread-id to saved pkey
+  std::map<pid_t, PKeyInfo> pid_key_map;
+  // MPK map safety mutex
+  std::mutex px;
```
Comment (ilovepi) [inline compiler-rt/lib/mpk_untrusted/alloc_site_handler.h]: what is the difference between this and mx? I cannot readily see from the comments. also, why are these `safety` mutexs? mutual exclusion is always a safety property. Is this an exclusive lock on pid_key_map? then what does mx lock?

File: compiler-rt/lib/mpk_untrusted/alloc_site_handler.h [hunk 7]
```diff
@@ -193,6 +206,27 @@ class AllocSiteHandler {
     }
   }
 
+  void storePidKey(pid_t threadID, PKeyInfo pkey) {
```
Comment (ilovepi) [inline compiler-rt/lib/mpk_untrusted/alloc_site_handler.h]: can we document what these APIs are for?

File: compiler-rt/lib/mpk_untrusted/alloc_site_handler.h [hunk 8]
```diff
@@ -94,6 +95,14 @@ class StatsTracker {
   }
 };
 
+class PKeyInfo {
```
Comment (ilovepi) [inline compiler-rt/lib/mpk_untrusted/alloc_site_handler.h]: how is this class to be used? If all fields are public, why not a struct?

File: compiler-rt/lib/mpk_untrusted/mpk_fault_handler.cpp [hunk 9]
```diff
@@ -90,7 +90,10 @@ void disableMPK(siginfo_t *si, void *arg) {
...
+  auto pid_key_info = handler->getPidKey(getpid());
```
Comment (ilovepi) [inline compiler-rt/lib/mpk_untrusted/mpk_fault_handler.cpp]: does this get a handler for a specific thread? also, doesn't `getpid()` return the process id and not the thread_id? i'm not sure what this does, given that `getpid()` should return the same integer for all threads. `sys/types.h` and `unistd.h` together define `gettid()` that will give you the thread id. only in single threaded applications will `getpid() == gettid()`.

**Review by ilovepi — CHANGES_REQUESTED** [review]
These need some associated tests. Thankfully, unit tests for the AllocSiteHandler should be simple, since it has no external state. 

if you need to test the getpid/gettid stuff, GTEST and GMOCK can probably give you enough to stub things out. Alternatively, you could define your own API for gettid that is only used when building tests, but this might not be very flexible in practice.

File: compiler-rt/lib/mpk_untrusted/alloc_site_handler.h (hunk 1 above)
Comment (mgdickerson) [reply compiler-rt/lib/mpk_untrusted/alloc_site_handler.h]: Changed to suggested name pkey_by_pid_map and renamed accessing functions in accordance to comments below.

File: compiler-rt/lib/mpk_untrusted/alloc_site_handler.h (hunk 2 above)
Comment (mgdickerson) [reply compiler-rt/lib/mpk_untrusted/alloc_site_handler.h]: Changed naming to better indicate data removal from map.

File: compiler-rt/lib/mpk_untrusted/mpk_fault_handler.cpp (hunk 3 above)
Comment (mgdickerson) [reply compiler-rt/lib/mpk_untrusted/mpk_fault_handler.cpp]: Added this portion of code to the `__mpk_untrusted` namespace.

File: compiler-rt/lib/mpk_untrusted/mpk_fault_handler.cpp (hunk 4 above)
Comment (mgdickerson) [reply compiler-rt/lib/mpk_untrusted/mpk_fault_handler.cpp]: Added a `REPORT` macro. Not sure if casing is important here (in terms of best practice), but I wanted to ensure I was distinct from the original Report function.

File: compiler-rt/lib/mpk_untrusted/mpk_fault_handler.cpp (hunk 5 above)
Comment (mgdickerson) [reply compiler-rt/lib/mpk_untrusted/mpk_fault_handler.cpp]: Changed naming to `getOrInit()` in accordance with comment.

File: compiler-rt/lib/mpk_untrusted/alloc_site_handler.h (hunk 6 above)
Comment (mgdickerson) [reply compiler-rt/lib/mpk_untrusted/alloc_site_handler.h]: Changed naming and comments to better indicate intended usage of both mutexes.

File: compiler-rt/lib/mpk_untrusted/alloc_site_handler.h (hunk 7 above)
Comment (mgdickerson) [reply compiler-rt/lib/mpk_untrusted/alloc_site_handler.h]: Added documentation notes.

File: compiler-rt/lib/mpk_untrusted/alloc_site_handler.h (hunk 8 above)
Comment (mgdickerson) [reply compiler-rt/lib/mpk_untrusted/alloc_site_handler.h]: Changed to a struct.

File: compiler-rt/lib/mpk_untrusted/mpk_fault_handler.cpp (hunk 9 above)
Comment (mgdickerson) [reply compiler-rt/lib/mpk_untrusted/mpk_fault_handler.cpp]: This call will get a shared reference to a singleton handler, then each access function ensures correct access patterns by using the mutex within the handler. Also switched to `gettid()` macro in accordance with suggestion, and added tests to ensure the macro functioned correctly (and as expected).

File: compiler-rt/lib/mpk_untrusted/mpk_fault_handler.cpp (hunk 4 above)
Comment (ilovepi) [reply compiler-rt/lib/mpk_untrusted/mpk_fault_handler.cpp]: All uppercase for preprocessor MACROS is normal

File: compiler-rt/lib/mpk_untrusted/alloc_site_handler.h (hunk 1 above)
Comment (ilovepi) [reply compiler-rt/lib/mpk_untrusted/alloc_site_handler.h]: sorry, I know I suggested it, but `pkey_by_pid_map` might not be great, since you're actually trying to track thread IDs. I made that comment before I saw that it was tracking tids.  i would suggest using `thread_id` instead of `tid`. `pid` is kind of ubiquitous, but `tid` can be confusing.

Another good option here is to define a typename for the mapping you want that is descriptive, and then the name of the map is less important, since its type has a sane name. 

this is up to you, but at minimum `pid` shouldn't be in the name, since you're actually tracking this per thread.

File: compiler-rt/lib/mpk_untrusted/mpk_fault_handler.h [hunk 10]
```diff
@@ -2,9 +2,19 @@
 #define MPKSEGFAULTHANDLER_H
 
 #include "mpk_common.h"
+#include <sys/syscall.h>
 #include <unistd.h>
 
+#ifndef SYS_gettid
```
Comment (ilovepi) [inline compiler-rt/lib/mpk_untrusted/mpk_fault_handler.h:13]: why is this necessary? libc should be linked in, so ... I'm confused as to thy you're issuing a system call directly and redefining a symbol that should be available. If there is a reason, then please document that

File: compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp [hunk 11]
```diff
@@ -69,9 +72,81 @@ TEST(getAllocSite, ValidBetweenAddress) {
...
+void *ThreadedGettid(void *tid) {
+  *(pid_t *)tid = gettid();
+  pthread_exit(NULL);
+}
+
+// Test gettid() macro:
+// 1) PID and TID are the same on single thread
+// 2) PID and TID are different in multi-thread
+TEST(gettid, MacroTest) {
+  EXPECT_EQ(getpid(), gettid());
```
Comment (ilovepi) [inline compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp]: doesn't this only work if you don't launch the test runner w/ multiple threads? IIRC GTest provides a way to run tests w/ process level isolation, but also in a multithreaded way similar to the default Rust test infrastructure.

File: compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp [hunk 12]
```diff
@@ -69,9 +72,81 @@ TEST(getAllocSite, ValidBetweenAddress) {
...
+
+#define NUM_THREADS 10
+
+void *setAndGetPKeyInfo(void *__unused) {
+  auto handle = AllocSiteHandler::getOrInit();
+  PKeyInfo pkinf(1, PKEY_DISABLE_ACCESS);
+  handle->storePKeyInfo(gettid(), pkinf);
+  EXPECT_EQ(pkinf.access_rights,
+            handle->popPendingPKeyInfo(gettid()).getValue().access_rights);
+  pthread_exit(NULL);
```
Comment (ilovepi) [inline compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp]: I notice here and some other places you're using `NULL`, but since this is C++ it's a bit more idiomatic to use `nullptr`.

File: compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp [hunk 13]
```diff
@@ -69,9 +72,81 @@ TEST(getAllocSite, ValidBetweenAddress) {
...
+  EXPECT_NE(getpid(), tid);
+}
+
+#define NUM_THREADS 10
+
+void *setAndGetPKeyInfo(void *__unused) {
+  auto handle = AllocSiteHandler::getOrInit();
+  PKeyInfo pkinf(1, PKEY_DISABLE_ACCESS);
+  handle->storePKeyInfo(gettid(), pkinf);
+  EXPECT_EQ(pkinf.access_rights,
```
Comment (ilovepi) [inline compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp:124]: does it make sense to do this check immediately? I would expect the check to happen after joining the threads. If the goal is to test contention, then using only 10 threads will likely not be enough.

File: compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp [hunk 14]
```diff
@@ -69,9 +72,81 @@ TEST(getAllocSite, ValidBetweenAddress) {
...
+
+  for (pthread_t &thread_id : threads) {
+    rc = pthread_create(&thread_id, NULL, setAndGetPKeyInfo, NULL);
+    if (rc)
+      FAIL() << "Error: Unable to create thread.\n";
+  }
+
+  for (auto thread_id : threads) {
+    rc = pthread_join(thread_id, &status);
+    if (rc)
```
Comment (ilovepi) [inline compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp]: Don't you need to check status too?

File: compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp [hunk 15]
```diff
@@ -69,9 +72,81 @@ TEST(getAllocSite, ValidBetweenAddress) {
...
-  auto ptr = (rust_ptr)(-1); // 0xffffffffffffffff
+  auto ptr = (rust_ptr)(-1); // max pointer size
   auto newAllocSite = std::make_shared<AllocSite>(ptr, sizeof(uint64_t), 1);
   EXPECT_FALSE(newAllocSite->containsPtr(++ptr));
 }
 
+TEST(faultingAllocs, addFaultAlloc) {
+  auto handle = AllocSiteHandler::getOrInit();
+  auto ptr = (rust_ptr)malloc(sizeof(uint64_t));
+  auto newAllocSite = std::make_shared<AllocSite>(ptr, sizeof(uint64_t), 1);
```
Comment (ilovepi) [inline compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp]: what is the 1 for?

File: compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp [hunk 16]
```diff
@@ -69,9 +72,81 @@ TEST(getAllocSite, ValidBetweenAddress) {
...
+  void *status;
+  rc = pthread_join(thread_id, &status);
+  if (rc) {
+    FAIL() << "Error: Unable to join thread.\n";
+  }
+
+  EXPECT_NE(getpid(), tid);
+}
+
+#define NUM_THREADS 10
```
Comment (ilovepi) [inline compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp]: why not just const or constexpr?  no need to use define in C++ unless you need to be c compatible

**Review by ilovepi — COMMENTED** [review]
mostly good, but I'd like to see some minor revisions to the test,and maybe some naming. Also, I'd like to see some more documentation of the design overall.  The documentation can be a separate PR if you want, but the longer this takes, the easier it becomes to lose track of the details.

File: compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp [hunk 17]
```diff
@@ -69,9 +72,81 @@ TEST(getAllocSite, ValidBetweenAddress) {
 }
 
 TEST(getAllocSite, PointerArithmeticOverflowWraps) {
-  auto ptr = (rust_ptr)(-1); // 0xffffffffffffffff
+  auto ptr = (rust_ptr)(-1); // max pointer size
```
Comment (ilovepi) [inline compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp]: you should probably derive ptr from a uintptr_t, since its you're casting from a signed integer here. you're only getting away w/ this because the compiler is being nice here, but it doesn't have to be. you might also assert that the value of ptr is the value you expect.

you can check the ptr width to determine the max numerical value., or just accept that the test will fail if the pointer size is less than 64-bits. We don't care about 32 bit platforms, but we do want to be sure that we don't succeed because we use a certain version of clang to compile our test code.

File: compiler-rt/lib/mpk_untrusted/alloc_site_handler.h (hunk 8 above)
Comment (ilovepi) [reply compiler-rt/lib/mpk_untrusted/alloc_site_handler.h]: can we document what the struct is intended to track?

File: compiler-rt/lib/mpk_untrusted/alloc_site_handler.h (hunk 2 above)
Comment (ilovepi) [reply compiler-rt/lib/mpk_untrusted/alloc_site_handler.h]: I'm not seeing how the new name better indicates that it should be removed from the map. Yes it uses 'pop', but it also is a bit odd to have `storePKeyInfo()` and `popPendingPkeyInfo()`.  If this is intended to be a stack, then maybe it should be a stack? if the `PKeyInfo` is only used to track a pending thing, then maybe it should be `PendingPkeyInfo`, or if you want to use `PKeyInfo` elsewhere, then maybe the code here should use something like this:

 ```C++
struct PendingPKeyInfo{ 
  PKeyInfo info;
}
```

Another point is that unlike address tracking, where you need to maintain ordering between elements, there is no need for thread association. A table lookup should be fine, and might perform significantly better. If you optimize this later, you may want to consider more optimized data structures for this type of lookup.

File: compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp (hunk 11 above)
Comment (ilovepi) [reply compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp]: also, since this is testing a property of linux, and not your code, I don't think you need to check this

File: compiler-rt/lib/mpk_untrusted/mpk_fault_handler.h (hunk 10 above)
Comment (mgdickerson) [reply compiler-rt/lib/mpk_untrusted/mpk_fault_handler.h:13]: Added notes for why this is necessary, as per our discussion on different versions of Glibc.

File: compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp (hunk 11 above)
Comment (mgdickerson) [reply compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp]: Removed macro test.

File: compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp (hunk 12 above)
Comment (mgdickerson) [reply compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp]: Switched to `nullptr` as recommended.

File: compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp (hunk 13 above)
Comment (mgdickerson) [reply compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp:124]: Removed test for now, still working on how to test contention properly.

File: compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp (hunk 14 above)
Comment (mgdickerson) [reply compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp]: Removed test requiring this check.

File: compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp (hunk 15 above)
Comment (mgdickerson) [reply compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp]: Switched the 1 for a constant named `PKEY_DEFAULT_VALUE` to indicate that it is just a default for testing.

File: compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp (hunk 16 above)
Comment (mgdickerson) [reply compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp]: Removed in addition with the test that was removed.

File: compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp (hunk 17 above)
Comment (mgdickerson) [reply compiler-rt/lib/mpk_untrusted/tests/unit/alloc_site_handler_tests.cpp]: swapped to deriving pointer from `uintptr_t`.

File: compiler-rt/lib/mpk_untrusted/alloc_site_handler.h (hunk 8 above)
Comment (mgdickerson) [reply compiler-rt/lib/mpk_untrusted/alloc_site_handler.h]: Added documentation for struct's intended usage.

File: compiler-rt/lib/mpk_untrusted/alloc_site_handler.h (hunk 1 above)
Comment (mgdickerson) [reply compiler-rt/lib/mpk_untrusted/alloc_site_handler.h]: Added a `thread_id` type name for the type of the map and then changed `pid` to `tid` for better clarity.

**Review by ilovepi — APPROVED** [review]
This is pretty good, but the tests will need some work. Also the preprocessor flags for stat tracking in the LLVM passes can become a compiler option instead. That can happen in a separate PR though
